test/functional/plugin/lsp_spec.lua

```diff
@@ -850,21 +851,43 @@ describe('LSP', function()
     end)
   end)
   describe('buf_diagnostics_save_positions', function()
-    it('stores the diagnostics in diagnostics_by_buf', function ()
+    it('stores the diagnostics in diagnostics_by_buf_and_client_id', function ()
       local diagnostics = {
         { range = {}; message = "diag1" },
         { range = {}; message = "diag2" },
       }
       exec_lua([[
-        vim.lsp.util.buf_diagnostics_save_positions(...)]], 0, diagnostics)
-      eq(1, exec_lua [[ return #vim.lsp.util.diagnostics_by_buf ]])
+        vim.lsp.util.buf_diagnostics_save_positions(...)]], 0, 1, diagnostics)
+      eq(1, exec_lua [[ return #vim.lsp.util.diagnostics_by_buf_and_client_id ]])
+      eq(1, exec_lua [[ return #vim.lsp.util.diagnostics_by_buf_and_client_id[1] ]])
       eq(diagnostics, exec_lua [[
-        for _, diagnostics in pairs(vim.lsp.util.diagnostics_by_buf) do
+        for _, diagnostics in pairs(vim.lsp.util.diagnostics_by_buf_and_client_id[1]) do
           return diagnostics
         end
       ]])
     end)
   end)
+  describe('buf_get_diagnostics', function()
+    it('returns diagnostics across client_ids', function()
```

Comment:
Would you add a test that updates the diagnostics for client B from the state where the diagnostics for client A and client B are displayed?